Implement nonbonded exclusions as summed potential term

mcwitt opened the pull request:
Note: work in progress

Currently exclusions are computed internally in the nonbonded potential ([here](https://github.com/proteneer/timemachine/blob/54068f0e7cdfe063f067f684d6f1dfa3947db996/timemachine/cpp/src/kernels/k_nonbonded.cuh#L758)). This PR will move computation of the exclusion terms into a separate potential, and recover the existing nonbonded potential as a sum (using `SummedPotential` added in #537). This will increase flexibility and modularity of the code.

mcwitt commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -0,0 +1,200 @@
+
+template <typename RealType>
+void __global__ k_nonbonded_pairs(
+    const int M, // number of pairs
```
Not sure if we have a convention for this already?

proteneer replied:
used to be called E, but I think anything works

mcwitt replied:
Makes sense; I had assumed we used `E` for "exclusions", and in that case we'd probably want something else for the general case

mcwitt commented on timemachine/cpp/src/kernels/k_nonbonded.cuh:
```diff
@@ -752,204 +752,3 @@ void __global__ k_nonbonded_unified(
     };
 }
-
-// tbd add restrict
```
Is this comment still relevant?

badisa replied:
I don't think so

proteneer commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -0,0 +1,202 @@
+#pragma once
```
We may want to add upfront that the order of associative operations in this file must be consistent with that of the k_nonbonded.cuh file in order for the exclusions to cancel out correctly.

mcwitt replied:
good call; done in 4a9d33d

badisa commented on timemachine/cpp/src/nonbonded_pairs.cu:
```diff
@@ -0,0 +1,143 @@
+    // construct identity permutation
+    if (Interpolated) {
+        gpuErrchk(cudaMalloc(&d_perm_, N_ * sizeof(*d_perm_)));
+        for (int i = 0; i < N_; i++) {
```
Can you set GPU memory this way? Pretty sure this needs to be done through code like this: https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/nonbonded.cu#L223-L229

mcwitt replied:
makes sense - I guess we could also allocate and initialize on the host and copy? (since this is only happening at construction, I don't think it'd be a bottleneck?)

badisa replied:
Allocating on the host would also work

mcwitt replied:
Should be fixed to allocate on host and copy as of e849535.

badisa commented on timemachine/cpp/src/nonbonded_pairs.cu:
```diff
@@ -0,0 +1,143 @@
+
+    const int tpb = 32;
+
+    dim3 dimGrid((M_ + tpb - 1) / tpb, 1, 1);
```
You can pass `(M_ + tpb - 1) / tpb` to kernel calls without having to construct dimGrids if you want a 1d set of computations

mcwitt replied:
Thanks, I think that makes sense. Changed in f4a07f0 (and also switched to using `ceil_divide` from `math_utils.h`).

maxentile commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -0,0 +1,208 @@
+        return;
+    }
+
+    int atom_i_idx = pair_idxs[pair_idx * 2 + 0];
+
+    RealType ci_x = coords[atom_i_idx * 3 + 0];
+    RealType ci_y = coords[atom_i_idx * 3 + 1];
+    RealType ci_z = coords[atom_i_idx * 3 + 2];
+    RealType ci_w = coords_w[atom_i_idx];
+
+    RealType dq_dl_i = dp_dl[atom_i_idx * 3 + 0];
+    RealType dsig_dl_i = dp_dl[atom_i_idx * 3 + 1];
+    RealType deps_dl_i = dp_dl[atom_i_idx * 3 + 2];
+    RealType dw_dl_i = dw_dl[atom_i_idx];
+
+    unsigned long long gi_x = 0;
+    unsigned long long gi_y = 0;
+    unsigned long long gi_z = 0;
+
+    int charge_param_idx_i = atom_i_idx * 3 + 0;
+    int lj_param_idx_sig_i = atom_i_idx * 3 + 1;
+    int lj_param_idx_eps_i = atom_i_idx * 3 + 2;
+
+    RealType qi = params[charge_param_idx_i];
+    RealType sig_i = params[lj_param_idx_sig_i];
+    RealType eps_i = params[lj_param_idx_eps_i];
+
+    unsigned long long g_qi = 0;
+    unsigned long long g_sigi = 0;
+    unsigned long long g_epsi = 0;
```
I've eyeballed for local self-consistency (e.g. that L42-68 is identical to L70-96, and that the lines to zip (`x`, `+ 0`), (`y`, `+ 1`), (`z`, `+ 2`) etc. seem right) and this all looks look fine!

Not required, but I find long stretches of manually unrolled index arithmetic tedious to verify (and sometimes they contain subtle typos that sneak past review). Would be a small quality-of-life improvement if we can find a more concise idiom for `.x`, `.y`, `.z` that doesn't introduce runtime cost...

mcwitt replied:
This makes sense to me. To be clear, the implementation here was copied mostly unmodified from `k_nonbonded.cuh`, so I don't think there would be any new bugs introduced in the index arithmetic. But I'm definitely in favor of finding a zero (or acceptable) cost abstraction to make correctness more evident.

mcwitt commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -0,0 +1,208 @@
+        g_qj -= FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DCHARGE>(charge_scale * qi * inv_dij * ebd);
+
+        real_du_dl -= delta_w * delta_prefactor * (dw_dl_i - dw_dl_j);
+        real_du_dl -= charge_scale * inv_dij * ebd * fix_nvidia_fmad(qj, dq_dl_i, qi, dq_dl_j);
```
Nice, I personally really like `strict_mul(a, b) + strict_mul(c, d)`. Maybe we could create an issue and implement this in a separate PR?

maxentile replied:
Migrated to https://github.com/proteneer/timemachine/issues/549 !

badisa commented on timemachine/cpp/src/nonbonded_pairs.cu:
```diff
@@ -0,0 +1,149 @@
+        gpuErrchk(cudaMemcpy(d_scales_, &scales[0], M_ * 2 * sizeof(*d_scales_), cudaMemcpyHostToDevice));
+
+        // construct identity permutation
+        int *perm = static_cast<int *>(malloc(N_ * sizeof(int)));
```
Since this is malloc'd don't we also need to free it? Using a std::vector<int> would avoid this

mcwitt replied:
Good call, thanks. I think using `std::vector` makes sense.

mcwitt replied:
Fixed in e849535.

mcwitt commented on timemachine/cpp/src/nonbonded_dense.cu:
```diff
@@ -408,37 +395,6 @@ void Nonbonded<RealType, Interpolated>::execute_device(
     }
 
-    // exclusions use the non-sorted version
-    if (E_ > 0) {
```
TODO: might want to add this check back in the appropriate place

mcwitt commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -0,0 +1,208 @@
+        gj_z -= FLOAT_TO_FIXED_NONBONDED(-delta_prefactor * delta_z);
+
+        // energy is size extensive so this may not be a good idea
+        energy -= FLOAT_TO_FIXED_NONBONDED(u);
```
TODO: the implementation here is still specialized to exclusions (returns the _negative_ sum of pair potentials). We'll probably want flip the convention and return the positive potential since we intend to use this as a generic nonbonded pairs potential.

mcwitt commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -0,0 +1,208 @@
+
+    unsigned long long energy = 0;
+
+    int is_vanilla =
```
TODO: more descriptive name for this?

badisa commented on timemachine/cpp/src/kernels/nonbonded_common.cuh:
```diff
@@ -0,0 +1,106 @@
+#pragma once
```
Since we now have a header and a source file, should we move the implementation over to the source file to have a consistent style?

mcwitt replied:
@badisa I might be mistaken but I think the templated functions and those declared with `__forceinline__` need to stay in the header file (templated functions because usage invokes template instantiation and `__forceinline__` so they can be inlined before optimization.

badisa commented on timemachine/cpp/src/nonbonded_pairs.cu:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    for (int i = 0; i < M_; i++) {
+        auto src = pair_idxs[i * 2 + 0];
```
@proteneer Do we use auto? I thought you had told me not to do that, but could be mistaken

mcwitt replied:
(This is not new; it was copied from existing code in `nonbonded.cu`.)

badisa commented on timemachine/cpp/src/nonbonded_pairs.cu:
```diff
@@ -0,0 +1,197 @@
+    const int num_tuples = Interpolated ? N * 2 : N;
+
+    for (int i = 0; i < num_tuples; i++) {
+        const int idx_charge = i * 3 + 0, idx_sig = i * 3 + 1, idx_eps = i * 3 + 2;
```
nit: Having this on three lines would be easier to read

mcwitt replied:
I like it, changed in 935428b

maxentile commented on timemachine/cpp/src/nonbonded.hpp:
```diff
@@ -1,102 +1,32 @@
-    jitify::KernelInstantiation compute_permute_interpolated_;
-    jitify::KernelInstantiation compute_add_du_dp_interpolated_;
+    NonbondedDense<RealType, Interpolated> dense_;
+    NonbondedPairs<RealType, true, Interpolated> exclusions_; // implement exclusions as negated NonbondedPairs
```
Subjective nit
```suggestion
    bool Negated = true;
    NonbondedPairs<RealType, Negated, Interpolated> exclusions_;
```

mcwitt replied:
I like it, added in c299d45. (needed `static const` for this to compile)

maxentile commented on timemachine/cpp/src/wrap_kernels.cpp:
```diff
@@ -902,9 +903,61 @@ template <typename RealType, bool Interpolated> void declare_nonbonded(py::modul
 template <typename RealType, bool Interpolated> void declare_nonbonded_pairs(py::module &m, const char *typestr) {
 
-    using Class = timemachine::NonbondedPairs<RealType, Interpolated>;
+    using Class = timemachine::NonbondedPairs<RealType, false, Interpolated>;
```
Nice -- I think this makes sense to keep the wrapper specialized to `Negated=false`

mcwitt replied:
Agreed -- I considered exposing this to the Python wrapper, but it seemed best to only wrap the non-negated version for now for simplicity.

maxentile commented on timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh:
```diff
@@ -154,27 +158,27 @@ void __global__ k_nonbonded_pairs(
             g_epsj += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DEPS>(-eps_grad * eps_i);
 
-            real_du_dl -= sig_grad * (dsig_dl_i + dsig_dl_j);
+            real_du_dl += sign * sig_grad * (dsig_dl_i + dsig_dl_j);
```
Difficult to verify at a glance if `accumulator -= x` --> `accumulator += sign * x` is applied everywhere it needs to be -- should `sign` also be applied to the `g_sigi` and `g_epsi` increments?

maxentile replied:
(Additionally, might apply `result = sign * accumulator` at the end, instead of to each of the increments)

mcwitt replied:
What do you think about factoring this into an inlined helper function as in 275d787?

mcwitt replied:
> Difficult to verify at a glance if accumulator -= x --> accumulator += sign * x is applied everywhere it needs to be

Agreed. I originally checked this by comparing the implementations in [`v_nonbonded_unified`](https://github.com/proteneer/timemachine/blob/669320ece66894d4204909f9359bc4e079de50cd/timemachine/cpp/src/kernels/k_nonbonded.cuh#L399) with [`k_nonbonded_exclusions`](https://github.com/proteneer/timemachine/blob/669320ece66894d4204909f9359bc4e079de50cd/timemachine/cpp/src/kernels/k_nonbonded.cuh#L747). But I think your idea of changing the sign at the final accumulation is more clearly correct.

mcwitt replied:
Oh, I was too hasty, 275d787 doesn't actually work. I'm not exactly sure why, but I'll have to think more carefully about the order of the floating-to-fixed conversions.

mcwitt replied:
> should sign also be applied to the g_sigi and g_epsi increments?

Good call, I missed these (they were accidentally correct for the `Negated = true` case, but we're currently missing a test for `Negated = false`). Fixed in 1ae5fa3.

maxentile replied:
>Good call, I missed these (they were accidentally correct for the Negated = true case, but we're currently missing a test for Negated = false). Fixed in 1ae5fa3.

Nice -- that looks good!

>What do you think about factoring this into an inlined helper function as in 275d787?
>idea of changing the sign at the final accumulation is more clearly correct.

Current implementation looks good! (Although I have a slight preference for flipping sign at the end rather than during accumulation (and reverting to multiplying by a sign constant rather than applying an inlined function).)

maxentile replied:
>I have a slight preference for flipping sign at the end rather than during accumulation

Ahh, @mcwitt points out an issue I had overlooked with this suggestion: the starting values could be nonzero!

maxentile reviewed (commented):
Looks great to me!

(I just want to stare a bit more at one stretch where sign-flip errors seem easy to make)

badisa commented on timemachine/cpp/src/wrap_kernels.cpp:
```diff
@@ -1010,6 +1124,18 @@ PYBIND11_MODULE(custom_ops, m) {
     declare_nonbonded<double, false>(m, "f64");
     declare_nonbonded<float, false>(m, "f32");
 
+    declare_nonbonded_dense<double, true>(m, "f64_interpolated");
```
We should add tests for the pairs and the dense, as this will be used but currently the wrappers aren't tested.

mcwitt replied:
Good call. Since this PR is already getting big, I'm thinking to revert exposing `NonbondedDense` and `NonbondedPairs` to user code for now (5d286a5). Then we can test them individually when they're exposed in a future PR.

mcwitt commented:
> lgtm pending mem-check/leak-check

Good call, I did find one issue, fixed in ab1edb8. There were also leaks detected in `master` when running `test_nonbonded`. I'll open a separate issue to deal with these.

```
========= LEAK SUMMARY: 7536 bytes leaked in 14 allocations
========= ERROR SUMMARY: 14 errors
```

mcwitt commented:
I added and reverted 38381b4 to demonstrate the test failure seen when implementing exclusions by passing negative scale parameters (this was the reason to switch to controlling negation with a template argument in fa0d85e).
